### ISSUE TITLE
unnecessary_cast false positive casting a Legacy type to a Nullable type.

### ISSUE THREAD

**Issue by jakemac53** [dart-lang/linter#2101]
When calling into a legacy API from an opted in api, the default type inference is to convert legacy types to their _non-nullable_ variants. When they may actually contain null, you need to explicitly specify that at the call site. 

Today the only way you can do it is by putting a type on the LHS of a declaration:

```
// assume `bar` is from some legacy library, and `baz` is of type `int*`
int? foo = bar.baz; 
```

If you attempt to do it with a cast to `int?` instead, the `unnecessary_cast` lint triggers, which is a false positive as this does affect behavior:

```
// triggers unnecessary cast lint :(
var foo = bar.baz as int?;
```

I propose that the unnecessary cast lint should thus never warn when casting a legacy type to a nullable type, as that does affect behavior.
